flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala

```diff
@@ -312,59 +323,125 @@ class JobManager(
 
       leaderSessionID = None
 
-    case RegisterTaskManager(
-      connectionInfo,
-      hardwareInformation,
-      numberOfSlots) =>
+    case msg: RegisterResourceManager =>
+      log.debug(s"Resource manager registration: $msg")
+
+      // ditch current resource manager (if any)
+      currentResourceManager = Option(msg.resourceManager())
+
+      val taskManagerResources = instanceManager.getAllRegisteredInstances.asScala.map(
+        instance => instance.getResourceId).toList.asJava
+
+      // confirm registration and send known task managers with their resource ids
+      sender ! decorateMessage(new RegisterResourceManagerSuccessful(self, taskManagerResources))
+
+    case msg: DisconnectResourceManager =>
+      log.debug(s"Resource manager disconnect: $msg")
+
+      currentResourceManager match {
+        case Some(rm) if rm.equals(msg.resourceManager()) =>
+          // we should ditch the current resource manager
+          log.debug(s"Disconnecting resource manager $rm.")
+          // send the old one a disconnect message
+          rm ! decorateMessage(new TriggerRegistrationAtJobManager(self))
```

Comment:
You're right that the message can be lost. Apart from the leader election service there is currently no mechanism to detect a lost connection from RM to the JM.

The above code is part of a special case where the RM doesn't reply to a TM registration and the JM decides to disconnect the RM. I suppose we should keep retrying to send the TriggerRegistrationAtJobManager until we receive a registration of a RM.